Add flower-client-authentication code example (#2999)
adap/flower: examples/flower-client-authentication/README.md

# Flower Client Authentication with PyTorch 🧪

> 🧪 = This example covers experimental features that might change in future versions of Flower
> Please consult the regular PyTorch code examples ([quickstart](https://github.com/adap/flower/tree/main/examples/quickstart-pytorch), [advanced](https://github.com/adap/flower/tree/main/examples/advanced-pytorch)) to learn how to use Flower with PyTorch.

> The following steps describe how to start a long-running Flower server (SuperLink) and then run a Flower App (consisting of a `ClientApp` and a `ServerApp`) with client authentication enabled.

Reviewer: ```suggestion
The following steps describe how to start a long-running Flower server (SuperLink) and a long-running Flower client (SuperNode) with client authentication enabled.
```

## Preconditions

Let's assume the following project structure:

```bash
$ tree .
.
├── client.py         # <-- contains `ClientApp`
├── server.py         # <-- contains `ServerApp`
├── task.py           # <-- task-specific code (model, data)
└── requirements.txt  # <-- dependencies
```

## Install dependencies

```bash
pip install -r requirements.txt

> ```

Reviewer: Can we move to the new example dependency management approach introduced by @chongshenng?

Reviewer 2: @danieljanes @danielnugraha I've modified this example to use the `hatchling` approach here #3377. Feel free to merge into this PR if you're happy with it!

## Generate public and private keys

```bash
bash ./generate.sh
```

> `generate.sh` is a script that generates per default three private and public keys, for server and two clients.
> You can generate more keys by specifying number of client credentials that you wish.
> Also, the script generates a csv file that includes each of the generated client credentials public keys.

Reviewer: ```suggestion
`generate.sh` is a script that (by default) generates three private and public key pairs for one server and two clients.
You can generate more keys by specifying the number of client credentials that you wish to generate.
The script also generates a CSV file that includes each of the generated (client) public keys.

Note that this script should only be used for development purposes and not for creating production key pairs.
```

```bash

> bash ./generate.sh {your_number_here}

Reviewer: ```suggestion
$ ./generate.sh {your_number_here}
```

```

## Start the long-running Flower server (SuperLink)

```bash
flower-superlink --insecure --require-client-authentication ./keys/client_public_keys.csv ./keys/server_credentials.pub ./keys/server_credentials
```

To start a long-running Flower server and enable client authentication is very easy, all you need to do is to type
`--require-client-authentication` followed by the path to the known `client_public_keys.csv`, server's public key

> `server_credentials.pub`, and server's private key `server_credentials`.

Reviewer: The order is wrong (it should be private/public instead of public/private)

## Start the long-running Flower client (SuperNode)

In a new terminal window, start the first long-running Flower client:

```bash

> flower-client-app client:app --insecure --authentication-keys ./keys/client_credentials_1.pub ./keys/client_credentials_1

Reviewer: Same as above re `--insecure`

Reviewer: Also: private key must go first

```

In yet another new terminal window, start the second long-running Flower client:

```bash

> flower-client-app client:app --insecure --authentication-keys ./keys/client_credentials_2.pub ./keys/client_credentials_2

Reviewer: Same as above re `--insecure`

Reviewer: Also: private key must go first

```

> If you generated more than 2 client credentials, you can add more clients by opening new terminal window and run the command
> above, don't forget to specify the correct client public and private keys for each client instance you created.

Reviewer: ```suggestion
If you generated more than 2 client credentials, you can add more clients by opening new terminal windows and running the command
above. Don't forget to specify the correct client private and public keys for each client instance you created.
```

## Run the Flower App

> With both the long-running server (SuperLink) and two clients (SuperNode) up and running, we can now run the actual Flower App:

Reviewer: ```suggestion
With both the long-running server (SuperLink) and two clients (SuperNode) up and running, we can now run the actual Flower ServerApp:
```

```bash
flower-server-app server:app --insecure
```

Or, to try the custom server function example, run:

```bash
flower-server-app server_custom:app --insecure

> ```

Reviewer: Let's remove this

# Flower Client Authentication with PyTorch 🧪

> 🧪 = This example covers experimental features that might change in future versions of Flower
> Please consult the regular PyTorch code examples ([quickstart](https://github.com/adap/flower/tree/main/examples/quickstart-pytorch), [advanced](https://github.com/adap/flower/tree/main/examples/advanced-pytorch)) to learn how to use Flower with PyTorch.

The following steps describe how to start a long-running Flower server (SuperLink) and a long-running Flower client (SuperNode) with client authentication enabled.

## Preconditions

> Let's assume the following project structure:

Reviewer 2: Let's walk them through the `git clone` process, like how we do with other examples:

Start by cloning the example project. We prepared a single-line command that you can copy into your shell which will checkout the example for you:
``` bash
git clone --depth=1 https://github.com/adap/flower.git _tmp && mv _tmp/examples/flower-client-authentication . && rm -rf _tmp && cd flower-client-authentication
```
This will create a new directory called `flower-client-authentication` with the following project structure:
